File: Leetcode/1143. Longest Common Subsequence.cpp
Problem link: https://leetcode.com/problems/longest-common-subsequence/
Problem Difficulty: Medium

class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        vector<vector<int>> dp(text1.length() + 1, vector<int> (text2.length() + 1));
        for (int i = 0; i < text1.length(); ++i) {
            for (int j = 0; j < text2.length(); ++j) {
                if (text1[i] == text2[j]) {
                    dp[i + 1][j + 1] = dp[i][j] + 1;
                }
                else {
                    dp[i + 1][j + 1] = max(dp[i][j + 1], dp[i + 1][j]);
                }
            }
        }
        return dp[text1.length()][text2.length()];
    }
};

File: Leetcode/1143. Longest Common Subsequence.java
// Problem Link: https://leetcode.com/problems/longest-common-subsequence/

class Solution {
    public int[][] dp;
    public int calculate(String s1, String s2, int i, int j) {
        if (i == s1.length() || j == s2.length()) return 0;

        if (s1.charAt(i) == s2.charAt(j)) {
            if (dp[i][j] < 0) {
                dp[i][j] = 1 + calculate(s1, s2, i + 1, j + 1);
            }  
            return dp[i][j];          
        } else {
            if (dp[i][j] < 0) {
                dp[i][j] = Math.max(calculate(s1, s2, i + 1, j), calculate(s1, s2, i, j + 1));
            }
            return dp[i][j];        
        }
    }

    public int longestCommonSubsequence(String text1, String text2) {
        dp = new int[text1.length()][text2.length()];
        for (int i = 0; i < dp.length; i++) {
            for (int j = 0; j < dp[i].length; j++) {
                dp[i][j]= -1;
            }
        }
        return calculate(text1, text2, 0, 0);
    }
}
